File: LeetCode (en comments)/912. Sort an Array.py
class Solution:
    def sortArray(self, sorting_list: List[int]) -> List[int]:
        if len(sorting_list) < 2:
            return sorting_list
        elif len(sorting_list) == 2:
            if sorting_list[0] > sorting_list[1]:
                return [sorting_list[1], sorting_list[0]]
            else:
                return sorting_list
        left = self.sortArray(sorting_list[:len(sorting_list) // 2])
        right = self.sortArray(sorting_list[len(sorting_list) // 2:])
        li, ri = 0, 0
        result = []
        while li < len(left) and ri < len(right):
            if left[li] < right[ri]:
                result.append(left[li])
                li += 1
            elif right[ri] < left[li]:
                result.append(right[ri])
                ri += 1
            else:
                result.append(left[li])
                result.append(right[ri])
                li += 1
                ri += 1
        while li < len(left):
            result.append(left[li])
            li += 1
        while ri < len(right):
            result.append(right[ri])
            ri += 1
        return result

File: LeetCode (en comments)/912. Sort an Array.rs
impl Solution {
    pub fn sort_array(nums: Vec<i32>) -> Vec<i32> {
        if nums.len() < 2 {
            return nums;
        } else if nums.len() == 2 {
            if nums[0] > nums[1] {
                return vec![nums[1], nums[0]];
            } else {
                return nums;
            }
        }
        let mut left = nums[0..nums.len() / 2].to_vec();
        let mut right = nums[nums.len() / 2..].to_vec();
        left = Solution::sort_array(left);
        right = Solution::sort_array(right);
        let (mut li, mut ri, mut i) = (0, 0, 0);
        let mut result = vec![0; nums.len()];
        while li < left.len() && ri < right.len() {
            if left[li] < right[ri] {
                result[i] = left[li];
                li += 1;
            } else if right[ri] < left[li] {
                result[i] = right[ri];
                ri += 1;
            } else {
                result[i] = left[li];
                i += 1;
                result[i] = right[ri];
                li += 1;
                ri += 1;
            }
            i += 1;
        }
        while li < left.len() {
            result[i] = left[li];
            i += 1;
            li += 1;
        }
        while ri < right.len() {
            result[i] = right[ri];
            i += 1;
            ri += 1;
        }
        result.resize(i, 0);
        result
    }
}
